### PR TITLE
blob: fix selected lines and text styles conflict

### PR DESCRIPTION
**Pull request by taras-yemets**
Closes https://github.com/sourcegraph/sourcegraph/issues/47364

1. Refactors selected lines highlighting to use CodeMirror `layer` extension instead of the line decorations.
With this approach, both selected lines and editor selection layers may be visible (with the latter taking precedence).
It makes selected text highlighted even if it is on a selected line. This approach is similar to how JetBrains IDE handles line breakpoint and text selection styles conflict (video attached to [this comment](https://github.com/sourcegraph/sourcegraph/pull/47378#issuecomment-1416079025)). 
2. Introduces another CSS variable: `--code-selection-bg-2`. It's used for user-selected text in the code editor. Its value is taken from my current VSCode theme as I couldn't find the suitable color in [`colors.scss`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/wildcard/src/global-styles/colors.scss?subtree=true).

Feedback on design (colors) and implementation is highly appreciated!

https://user-images.githubusercontent.com/25318659/216649664-494d17d9-f930-4588-a337-2b307c58894f.mov

## Test plan
Tested manually (video attached).

## App preview:

- [Web](https://sg-web-taras-yemets-cm-blob-fix-selected.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by taras-yemets** [comment]
> This approach is similar to how JetBrains IDE handles line breakpoint and text selection styles conflict (video in the first comment).


https://user-images.githubusercontent.com/25318659/216650789-1eb1b44d-ce41-4b21-85fb-23610968f410.mov

File: client/web/src/repo/blob/codemirror/code-folding.tsx [hunk 1]
```diff
@@ -126,6 +126,7 @@ export function codeFoldingExtension(): Extension {
         foldGutter({
             markerDOM(open: boolean): HTMLElement {
                 const container = document.createElement('div')
+                container.classList.add('fold-marker')
```
Comment (philipp-spiess) [inline client/web/src/repo/blob/codemirror/code-folding.tsx:129]: Is this an unrelated fix? Not sure I understand this in context of the selection highlight

**Review by philipp-spiess — APPROVED** [review]
Great fix!

**Comment by fkling** [comment]
Sorry, I haven't had time to look at the changes, but I just wanted to mention that one of the examples mentions the following:

> Colors can be adjusted throughout, but when adding background colors to content, it is recommended to use partially transparent colors. That way such a style doesn't hide other styling behind it (including the selection).

https://codemirror.net/examples/styling/

Again, haven't looked into this problem here specifically but if using a transparent color would be a simpler fix we should consider that too.

Feel free to ignore :)

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (+0.12 kb) | 0.01% (+0.80 kb) | 0.01% (+0.68 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/taras-yemets/cm-blob-fix-selected-lines-and-text-styles/compare-report.html;tab=live_object) for a full comparison between the commits a96415e and 7e55a5b or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: client/web/src/repo/blob/codemirror/code-folding.tsx (hunk 1 above)
Comment (taras-yemets) [reply client/web/src/repo/blob/codemirror/code-folding.tsx:129]: The fold icon container needs fixed width to avoid content shifts when the react component (icon) is rendered. Previously we set this width on the gutter column. With [the change introduced by this PR](https://github.com/sourcegraph/sourcegraph/pull/47378/files#diff-9eedf1330af1bfd48a4db4a147fed560a0cff4fffbbcdadd8afc89b1948a8787R140-R154), we add extra padding to last gutter's `.cm-gutterElement` (child of of `.cm-gutter` and parent of `.fold-marker` in this case). To ensure that the fold icon container width is unaffected, we move `width` rule one level down. That's why we add this extra wrapper here.

**Comment by taras-yemets** [comment]
>Colors can be adjusted throughout, but when adding background colors to content, it is recommended to use partially transparent colors. That way such a style doesn't hide other styling behind it (including the selection).
>
>https://codemirror.net/examples/styling/
>
>Again, haven't looked into this problem here specifically but if using a transparent color would be a simpler fix we should consider that too.

Nice idea! Thank you, @fkling!
Implementation with transparent selected line background would be much simpler (we just don't need that [extra layer](https://github.com/sourcegraph/sourcegraph/pull/47378/files#diff-9eedf1330af1bfd48a4db4a147fed560a0cff4fffbbcdadd8afc89b1948a8787R89-R129)). 

My main concern here is design - with this approach, we end up having three different line backgrounds. It may be confusing for users.

Screenshot 2023-02-06 at 08 04 33

**Comment by fkling** [comment]
Totally, I just wanted to mention it 🙂
